### PR TITLE
List only secret for gpg options that manage secret keys

### PR DESCRIPTION
**Pull request by spenserblack**
## Description

This modifies the internal `__fish_complete_gpg_user_id` to take an optional argument for the "key type" to modify `--list-keys` to something like `--list-secret-keys`. This is then used to list only secret keys with `gpg --export-secret-keys` completions. There are probably more uses where only secret keys should be listed, but this is what I found so far. I'll keep this as a draft for now in case I should limit other completions to only the secret keys.

## TODOs:

- [ ] Changes to fish usage are reflected in user documentation/manpages.
- [ ] Tests have been added for regressions fixed
- [ ] User-visible changes noted in CHANGELOG.rst

******

## TODOs (for this PR specifically)

- [x] `--delete-secret-key[s]`
- [x] `--try-secret-key`
- [x] `--list-secret-keys` (getting a list of secret keys to complete `--list-secret-keys` feels a little redundant :laughing:)
- [x] Also update `__fish_complete_gpg_key_id`?
- [x] Fix completions for secret key IDs, specifically `--try-secret-key` (the `pub*` case is never hit for secret keys)

### CODE REVIEW

**Comment by krobelus** [comment]
> There are probably more uses where only secret keys should be listed, but this is what I found so far. I'll keep this as a draft for now in case I should limit other completions to only the secret keys.

Ok, can you check?

File: share/functions/__fish_complete_gpg_user_id.fish [hunk 1]
```diff
@@ -1,11 +1,11 @@
...
-function __fish_complete_gpg_user_id -d "Complete using gpg user ids" -a __fish_complete_gpg_command
+function __fish_complete_gpg_user_id -d "Complete using gpg user ids" -a __fish_complete_gpg_command __key_type
     # gpg doesn't seem to like it when you use the whole key name as a
     # completion, so we skip the <EMAIL> part and use it as a description.
     # It also replaces \x3a from gpg's output with colons.
     #
     # TODO: I tried with <EMAIL> and it worked, this was possibly fixed in gpg.
     # Regardless, it's probably nicer as a description.
-    $__fish_complete_gpg_command --list-keys --with-colon | string split -a -f 10 : | string replace '\x3a' : | string replace -r '(.*) <(.*)>' '$1\t$2'
+    $__fish_complete_gpg_command (string join -- - --list $__key_type keys) --with-colon | string split -a -f 10 : | string replace '\x3a' : | string replace -r '(.*) <(.*)>' '$1\t$2'
```
[the remark this answers is not in the document]
Comment (spenserblack) [reply share/functions/__fish_complete_gpg_user_id.fish:10]: I was thinking that this parameter could eventually also be used to build `--list-public-keys` for other completions.

Though I agree that it's easier if the param is set in a separate line than joined inline. Wasn't sure what the preferred style was.

**Comment by spenserblack** [comment]
>Ok, can you check?

Yeah, no problem. I just want to resolve this conversation https://github.com/fish-shell/fish-shell/pull/8712#discussion_r804464412 before I change the other completions.

Changed title to reflect purpose of PR.

Even though this PR is still a WIP, I guess I should mark it "ready for review" since it's being reviewed :laughing:

**Comment by spenserblack** [comment]
This commit history is quite messy. Please let me know if you want me to rebase, and if `--autosquash` would be good enough.

**Comment by krobelus** [comment]
> This commit history is quite messy. Please let me know if you want me to rebase (e.g. if you're not going to squash merge this), and if `--autosquash` would be good enough.

history after --autosquash looked alright but squashing everything feels a bit better, so I did that, merged as 63d7386a36aa3a0fe40ef0fe0c1c1f1c1e9b73e5.
(BTW I think force-pushing is fine; it breaks github but there are other ways to explain what changed, if necessary)

**Comment by spenserblack** [comment]
Oh yeah, I was happy to force-push myself, I just didn't want to start rewriting the history until it was confirmed no additional changes were wanted/needed 🙂

**Pull request closed without merging**